[RISCV] Improve performCONCAT_VECTORCombine stride matching

michaelmaitland opened the pull request:
I intend to commit the two commits in this patch independently:

    Pre-commit concat-vectors-constant-stride tests.ll

    This patch commits tests that can be optimized by improving
    performCONCAT_VECTORCombine to do a better job at decomposing the base
    pointer and recognizing a constant offset.

    Improve performCONCAT_VECTORCombine stride matching

    If the load ptrs can be decomposed into a common (Base + Index) with a
    common constant stride, then return the constant stride. This matcher
    enables some additional optimization since BaseIndexOffset is capable of
    decomposing the load ptrs to (add (add Base, Index), Stride) instead of
    (add LastPtr, Stride) or (add NextPtr, Stride) that matchForwardStrided and
    matchReverseStrided use, respectively.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -13821,6 +13822,58 @@ static SDValue performCONCAT_VECTORSCombine(SDNode *N, SelectionDAG &DAG,
+
+    // Check that constraints hold for all subsequent loads and the ConstStride
+    // is the same.
+    for (auto Idx : enumerate(Loads.drop_front(2))) {
```
Do we need to use enumerate here? I don't think the index is used.

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -13821,6 +13822,58 @@ static SDValue performCONCAT_VECTORSCombine(SDNode *N, SelectionDAG &DAG,
+      if (!AllValidOffset)
+        return SDValue();
+      BaseIndexMatch |= BaseLdBIO.equalBaseIndex(BIO, DAG);
+      // Add 3 to index because the first two loads have been processed before
```
Is this comment stale? I don't see any `3`

topperc commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -13862,13 +13915,21 @@ static SDValue performCONCAT_VECTORSCombine(SDNode *N, SelectionDAG &DAG,
+    } else {
+      Stride = matchConstantStride(N->ops());
+      if (Stride) {
+        Reversed = cast<ConstantSDNode>(Stride)->getSExtValue() < 0;
```
Is this redundant with the only remaining use of `Reversed`?

michaelmaitland replied:
There is a use of Reversed below.

preames commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -13862,13 +13913,21 @@ static SDValue performCONCAT_VECTORSCombine(SDNode *N, SelectionDAG &DAG,
-      return SDValue();
+    if (Stride) {
+      Reversed = true;
+      Stride = DAG.getNegative(Stride, DL, Stride->getValueType(0));
```
The original code deferred forming the negate so that we didn't form the node if were going to bail out before completing the transform.

preames commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -13821,6 +13822,56 @@ static SDValue performCONCAT_VECTORSCombine(SDNode *N, SelectionDAG &DAG,
+    for (auto &Use : Loads.drop_front(2)) {
+      auto *Ld = cast<LoadSDNode>(Use);
+      BaseIndexOffset BIO = BaseIndexOffset::match(Ld, DAG);
+      AllValidOffset &= BIO.hasValidOffset();
```
You don't need to update these variables as you're already doing early return.  Reaching the end of the loop implies these must be true.

preames commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -13821,6 +13822,56 @@ static SDValue performCONCAT_VECTORSCombine(SDNode *N, SelectionDAG &DAG,
+    // loads.
+    if (Loads.size() < 2)
+      return SDValue();
+    BaseIndexOffset BaseLdBIO =
```
Hm, an idea...

Looking at this code, I realized we could factor out a helper routine of the form optional<int64_t> getPtrDiff(LD1, LD2).  Then we could loop over all of the indices, and check that the offset is known and the same.

After that, it occurred to me that both of the previous two matchers can *also* be written as rules inside a getPtrDiff helper.  (The return value has to become SDValue instead, but otherwise it works.)

Maybe we should switch over in a separate commit, and then implement this matcher in that style?

michaelmaitland replied:
> I realized we could factor out a helper routine of the form optional<int64_t> getPtrDiff(LD1, LD2). Then we could loop over all of the indices, and check that the offset is known and the same.

This seems to be very close to what BaseIndexOffset match does since if the (Base+Index) matches then the ptr diff is Offset. We could definitely add this helper method to simplify our checks in matchConstantStride.

 > After that, it occurred to me that both of the previous two matchers can also be written as rules inside a getPtrDiff helper. (The return value has to become SDValue instead, but otherwise it works.)

What happens in the case where there is no constant stride in matchFoward and matchReverse (i.e. stride can be a common add SDValue that is not a SDConstant).

michaelmaitland replied:
I think I understand, `getPtrDiff` would return the canonicalized representation of the non-constant value. We wouldn't be able to use BaseIndexOffset here. Do you have any ideas on what we could use?

lukel97 replied:
I pulled out getPtrDiff myself when trying wrap my head around this, hope this doesn't step on anyones toes: #69068
We should be able to extend it to check BaseIndexOffsets and return `DAG.getConstant` like we do here if they match

michaelmaitland replied:
@lukel97 thanks for your refactor. This change is very simple with getPtrDiff. I've updated this patch to be stacked on top of https://github.com/llvm/llvm-project/pull/69068

michaelmaitland committed:
[RISCV] Pre-commit concat-vectors-constant-stride.ll

This patch commits tests that can be optimized by improving
performCONCAT_VECTORCombine to do a better job at decomposing the base
pointer and recognizing a constant offset.

michaelmaitland committed:
[RISCV] Improve performCONCAT_VECTORCombine stride matching

If the load ptrs can be decomposed into a common (Base + Index) with a
common constant stride, then return the constant stride.

michaelmaitland commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -13798,60 +13798,48 @@ static SDValue performCONCAT_VECTORSCombine(SDNode *N, SelectionDAG &DAG,
+    BaseIndexOffset BIO1 = BaseIndexOffset::match(Ld1, DAG);
+    BaseIndexOffset BIO2 = BaseIndexOffset::match(Ld2, DAG);
+    if (BIO1.equalBaseIndex(BIO2, DAG))
+      return {{DAG.getConstant(BIO2.getOffset() - BIO1.getOffset(), DL,
```
You are right. Let me take a look.

michaelmaitland replied:
I think what you propose makes sense. I thought about trying to do it using std::function to defer creation of constant or negative, but that doesn't allow us to compare `getPtrDiff() != BaseDiff`. We need to do a comparison by value on three cases:

1. SDValue, true
2. SDValue, false
3. int64_t, false

And the pair<variant, bool> structure permits us to do this check.

michaelmaitland committed:
Defer construction of constant SDNode

lukel97 reviewed (approved):
LGTM. Just leaving a note here that I think we'll fail to detect sequences of pointers that are a mixture of (base + index) and (add LastPtr, constStride), because the latter will be an SDValue and not a int64_t. But it sounds like an unlikely pattern to show up and I doubt its worth the complexity

michaelmaitland commented:
> LGTM. Just leaving a note here that I think we'll fail to detect sequences of pointers that are a mixture of (base + index) and (add LastPtr, constStride), because the latter will be an SDValue and not a int64_t. But it sounds like an unlikely pattern to show up and I doubt its worth the complexity

My guess is that if BaseIndexOffset couldn't resolve to a match, then we would have no way to compare a constant to a non-constant anyway.

michaelmaitland commented:
Committed in 30ca258614dd231e23f45ad1188905acadb86e66 and c319c741463a039c2323825b149df70cbe535c67 to preserve pre-commit diff.
